### PR TITLE
feat: collect stats from heartbeats

### PR DESCRIPTION
**Pull request by fengjiachun**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add two handlers for handling heartbeats, one for parsing and caching heartbeat data, and another for persisting the data so that follower nodes can see it after a certain delay.
The persisting data handler will not be completed in this PR.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/meta-srv/src/handler/collect_stats_handler.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+use dashmap::DashMap;
+
+use super::node_stat::Stat;
+use crate::error::Result;
+use crate::handler::{HeartbeatAccumulator, HeartbeatHandler};
+use crate::metasrv::Context;
+
+pub struct CollectStatsHandler {
+    max_size_per_key: usize,
+    cache: DashMap<(u64, u64), VecDeque<Stat>>,
```
Comment (MichaelScofield) [inline src/meta-srv/src/handler/collect_stats_handler.rs:29]: better make a type alias of `(u64, u64)`

File: src/meta-srv/src/handler/collect_stats_handler.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+
+use super::node_stat::Stat;
+use crate::error::Result;
+use crate::handler::{HeartbeatAccumulator, HeartbeatHandler};
+use crate::metasrv::Context;
+
+pub struct CollectStatsHandler {
+    max_size_per_key: usize,
```
Comment (MichaelScofield) [inline src/meta-srv/src/handler/collect_stats_handler.rs:28]: "size" is meaningless here

```suggestion
    max_cached_stats_per_key: usize,
```

File: src/meta-srv/src/handler/collect_stats_handler.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                            acc.stats = deque.drain(..).collect();
+                        }
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(VecDeque::from([stat]));
+                    }
+                }
+            }
+            Err(_) => {
+                debug!("Incomplete heartbeat data: {:?}", req);
```
Comment (MichaelScofield) [inline src/meta-srv/src/handler/collect_stats_handler.rs:76]:
```suggestion
            Err(e) => {
                error!(e; "Incomplete heartbeat data: {req:?}");
```

File: src/meta-srv/src/handler/node_stat.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    /// The write capacity units during this period
+    pub wcus: u64,
+    /// Approximate bytes of this region
+    pub approximate_bytes: u64,
+    /// Approximate number of rows in this region
+    pub approximate_rows: u64,
+}
+
+impl TryFrom<&HeartbeatRequest> for Stat {
+    type Error = ();
```
Comment (MichaelScofield) [inline src/meta-srv/src/handler/node_stat.rs:61]: `()` does not provide any useful info, at least:

```suggestion
    type Error = String;
```

File: src/meta-srv/src/handler/persist_stats_handler.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        &self,
+        _req: &HeartbeatRequest,
+        ctx: &Context,
+        acc: &mut HeartbeatAccumulator,
+    ) -> Result<()> {
+        if ctx.is_skip_all() || acc.stats.is_empty() {
+            return Ok(());
+        }
+
+        // TODO: remove stats from `acc` and persist to store
```
Comment (MichaelScofield) [inline src/meta-srv/src/handler/persist_stats_handler.rs:35]: please follow our todo convention, add your name here

File: src/meta-srv/src/handler/node_stat.rs (hunk 4 above)
Comment (fengjiachun) [reply src/meta-srv/src/handler/node_stat.rs:61]: Don't need error info, just need to know if it failed

File: src/meta-srv/src/handler/collect_stats_handler.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                match self.cache.entry(key) {
+                    Entry::Occupied(mut e) => {
+                        let deque = e.get_mut();
+                        deque.push_front(stat);
+                        if deque.len() >= self.max_cached_stats_per_key {
+                            acc.stats = deque.drain(..).collect();
+                        }
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(VecDeque::from([stat]));
```
Comment (Fengys123) [inline src/meta-srv/src/handler/collect_stats_handler.rs:73]: nit: use VecDeque::with_capacity(10) to reduce memory allocation
```suggestion
                        let mut stat_vec = VecDeque::with_capacity(10);
                        stat_vec.push_front(stat);
                        e.insert(stat_vec);

```

**Review by Fengys123 — APPROVED** [review]
lgtm

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**